### PR TITLE
issue/3168: Fixes to config inheritance, added _showEndOfPage

### PR DESCRIPTION
**Pull request by oliverfoster**
fixes https://github.com/adaptlearning/adapt_framework/issues/3168

### Changed
* Simplified config derivations and fixed bugs
* Simplified locking and component additions, each model will apply its own config as derived rather than nested loops from ancestors
* Renamed some variables, added some comments

### Added
* `_showEndOfPage` so that the final button in the last article can be hidden / shown

### Fixed
* Can now go to next page on end (part fix for https://github.com/adaptlearning/adapt_framework/issues/1568)

### CODE REVIEW

File: js/TrickleButtonModel.js [hunk 1]
```diff
@@ -72,20 +72,26 @@ export default class TrickleButtonModel extends ComponentModel {
...
   calculateButtonText() {
-    const siteId = this.get('_trickleConfigId');
-    const trickleConfig = getModelConfig(Adapt.findById(siteId));
+    const trickleConfig = getModelConfig(this.getParent());
     let isStart = false;
     let isFinal = false;
     if (trickleConfig._onChildren) {
       const parentId = this.getParent().get('_id');
-      const trickleParent = Adapt.findById(siteId);
+      const trickleParent = getModelContainer(this.getParent());
```
Comment (eleanor-heath) [inline js/TrickleButtonModel.js:92]: can you make a const out of this.getParent() to use here and on line 87.

File: js/models.js [hunk 2]
```diff
@@ -83,46 +88,60 @@ export function setModelDefaults(model) {
...
- * @param {object} config
+ * @returns {Object}
  */
-export function setModelConfig(model, config) {
-  model.unset('_trickleConfigId');
-  model.set('_trickle', config);
+export function getModelConfig(model) {
+  const inheritance = getModelInheritanceChain(model);
+  if (!inheritance?.length) return null;
+  if (isModelArticleWithOnChildren(model)) return null;
```
Comment (eleanor-heath) [inline js/models.js:123]: could combine these 2 if statements to return null.
`if (!inheritance?.length || isModelArticleWithOnChildren(model)) return null;`

File: js/models.js [hunk 3]
```diff
@@ -216,21 +211,21 @@ export function applyLocks() {
     if (wasLocked === isLocked) return;
     model.set('_isLocked', isLocked);
   });
+  log();
```
Comment (eleanor-heath) [inline js/models.js:214]: empty log?

File: js/models.js [hunk 4]
```diff
@@ -244,45 +239,52 @@ export function _getSelfAndAncestorNextSiblings(fromModel) {
...
-        _renderPosition: 'outer-append'
-      });
-      trickleModel.setupModel();
-      // This line would usually append a trickle button to an article or block
-      buttonModelSite.getChildren().add(trickleModel);
+  data.forEach(buttonModelSite => {
+    if (buttonModelSite instanceof CourseModel) return;
+    let trickleConfig = getModelConfig(buttonModelSite);
+    if (!trickleConfig || !trickleConfig?._isEnabled) return;
+    if (buttonModelSite.get('_isTrickleSiteConfigured')) return;
```
Comment (eleanor-heath) [inline js/models.js:257]: can combine the if statements here on line 256 and 257

**Review by eleanor-heath — APPROVED** [review]

File: js/models.js (hunk 3 above)
Comment (oliverfoster) [reply js/models.js:214]: It calls the function beneath is. I'll change the name.

**Review by tomgreenfield — APPROVED** [review]

**Pull request merged**